pkg/ccl/crosscluster/logical/lww_kv_processor_test.go

```diff
@@ -0,0 +1,175 @@
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package logical
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/ccl/crosscluster/replicationtestutils"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descs"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/desctestutils"
+	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/stretchr/testify/require"
+)
+
+func TestKVWriterUpdateEncoding(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	ctx := context.Background()
+
+	srv, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{})
+	defer srv.Stopper().Stop(ctx)
+	s := srv.ApplicationLayer()
+
+	runner := sqlutils.MakeSQLRunner(sqlDB)
+
+	tableWithNullableColumns := `CREATE TABLE %s (pk INT PRIMARY KEY, payload1 STRING NULL, payload2 STRING NULL)`
+
+	tableNumber := 0
+	createTable := func(t *testing.T, schema string) string {
+		tableName := fmt.Sprintf("tab%d", tableNumber)
+		runner.Exec(t, fmt.Sprintf(schema, tableName))
+		runner.Exec(t, fmt.Sprintf(
+			"ALTER TABLE %s "+lwwColumnAdd,
+			tableName))
+		tableNumber++
+		return tableName
+	}
+
+	type encoderFn func(datums ...interface{}) roachpb.KeyValue
+
+	setup := func(t *testing.T, schema string) (string, RowProcessor, encoderFn) {
+		tableName := createTable(t, schema)
+		desc := desctestutils.TestingGetPublicTableDescriptor(s.DB(), s.Codec(), "defaultdb", tableName)
+
+		rp, err := newKVRowProcessor(ctx,
+			&execinfra.ServerConfig{
+				DB:           s.InternalDB().(descs.DB),
+				LeaseManager: s.LeaseManager(),
+			}, &eval.Context{
+				Codec:    s.Codec(),
+				Settings: s.ClusterSettings(),
+			}, map[descpb.ID]sqlProcessorTableConfig{
+				desc.GetID(): {
+					srcDesc: desc,
+				},
+			})
+		require.NoError(t, err)
+		return tableName, rp, func(datums ...interface{}) roachpb.KeyValue {
+			kv := replicationtestutils.EncodeKV(t, s.Codec(), desc, datums...)
+			kv.Value.Timestamp = hlc.Timestamp{WallTime: timeutil.Now().UnixNano()}
+			return kv
+		}
+	}
+
+	insertRow := func(rp RowProcessor, keyValue roachpb.KeyValue, prevValue roachpb.Value) error {
+		_, err := rp.ProcessRow(ctx, nil, keyValue, prevValue)
+		return err
+	}
+
+	t.Run("one-NULL-to-not-NULL", func(t *testing.T) {
```

Comment:
for all these test cases, if we randomly send a stale previous value, everything should just work, right? I.e. we could randomly choose to send kv1 instead of kv2 as the previous value.

Reply:
Yup, added some code that uses a random previous value about half of the time.